### PR TITLE
Update Go version to 1.17

### PR DESCRIPTION
**Pull request by zrhoffman**
This PR
- Makes the *Is it time to update Go?* GHA workflow update `GO_VERSION` before setting up the Go version specified in `GO_VERSION`
- Makes the *build-rpms* GHA build the ATC builder images if the GO_VERSION has changed in the scope of the PR, or in the last commit if it's not being run on a PR
- Updates the Go version to 1.17 and updates the `golang.org/x/` dependencies.

See the Go 1.17 [release notes](https://golang.org/doc/devel/release#go1.17) for details.
<hr>

## Which Traffic Control components are affected by this PR?

- CDN in a Box - Enroller
- Grove
- Traffic Control Client (Go)
- Traffic Monitor
- Traffic Ops
- Traffic Ops ORT
- Traffic Stats
- CI tests for Go components

## What is the best way to verify this PR?

Run unit tests and API tests

## PR submission checklist
- [x] Existing tests are sufficient 
- [x] The documentation pulls the Go version from `GO_VERSION`, no other documentation changes are necessary
- [x] This PR has a CHANGELOG.md entry 
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://apache.org/security) for details)

### CODE REVIEW

**Comment by zrhoffman** [comment]
Rebased to fix the changelog merge conflict

File: .github/actions/pr-to-update-go/pr_to_update_go/__main__.py [hunk 1]
```diff
@@ -20,13 +21,17 @@
 
 
 def main() -> None:
+	parser = ArgumentParser()
+	parser.add_argument('--update_version_only', type=bool, default=False)
```
[the remark this answers is not in the document]
Comment (zrhoffman) [reply .github/actions/pr-to-update-go/pr_to_update_go/__main__.py]: Using kebab case in 3795649b16
Added help text in d4571fe00e

File: .github/actions/pr-to-update-go/pr_to_update_go/go_pr_maker.py [hunk 2]
```diff
@@ -128,13 +129,23 @@ def run(self) -> None:
 			print(f'Go version is up-to-date on {target_branch}, nothing to do.')
 			return
 
+		commit: Union[Commit, None] = None
```
[the remark this answers is not in the document]
Comment (zrhoffman) [reply .github/actions/pr-to-update-go/pr_to_update_go/go_pr_maker.py]: Using `Optional[X]` in 92641a7e6f

**Comment by ocket8888** [comment]
t3c failure looks related to this PR and not spurious:
```
go: go.mod file not found in current directory or any parent directory; see 'go help modules'
```

**Comment by zrhoffman** [comment]
> t3c failure looks related to this PR and not spurious:
> 
> ```
> go: go.mod file not found in current directory or any parent directory; see 'go help modules'
> ```

That error always happened because `go mod vendor` was run before `cd`ing into the ATC directory.  Once in the ATC directory, `go get ./...` was run instead (changed to `go mod vendor` in 516df79ddc).

As reported in golang/go#45609, `go mod` reports no dependencies in symlinked directories, resulting in `go mod vendor` emptying out the vendor directory. Resolving the symlink when `cd`ing into the ATC directory in c931a5f2e3.

**Comment by zrhoffman** [comment]
Okay, the logic for whether to build or pull the builder images is working.

**Comment by zrhoffman** [comment]
Now it's *actually* working.

File: .github/actions/build-rpms/build-rpms.sh [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	files_changed="$(curl "${GITHUB_API_URL}/repos/${GITHUB_REPOSITORY}/pulls/${pr_number}/files" | jq -r .[].filename)"
+else
+	files_changed="$(git diff-tree --no-commit-id --name-only -r "$GITHUB_SHA")"
+fi
+if <<<"$files_changed" grep '^GO_VERSION$'; then
+	pkg_command+=(-b)
+fi
+
+if [[ -z "${ATC_COMPONENT:-}" ]]; then
+	echo 'Missing environment variable ATC_COMPONENT' >/dev/stderr
```
Comment (ocket8888) [inline .github/actions/build-rpms/build-rpms.sh:38]: does piping to fd 2 not work? I don't think it's actually a problem, I just don't normally see it done this way, so I was curious.

File: .github/actions/pr-to-update-go/pr_to_update_go/go_pr_maker.py [hunk 4]
```diff
@@ -316,18 +327,18 @@ def set_go_version(self, go_version: str, commit_message: str,
...
 		return commit
 
-	def update_golang_org_x(self, previous_commit: Commit) -> Union[GitCommit, None]:
+	def update_golang_org_x(self, previous_commit: Commit) -> Optional[GitCommit]:
 		"""
 		:param previous_commit:
 		:type previous_commit:
 		:return:
-		:rtype: Union[GitCommit, None]
+		:rtype: Optional[GitCommit]
```
Comment (ocket8888) [inline .github/actions/pr-to-update-go/pr_to_update_go/go_pr_maker.py:335]: fwiw, I wouldn't bother adding return and parameter types like this when that information is already available in the annotations.

**Review by ocket8888 — APPROVED** [review]
This all LGTM - I'll leave it open for a bit in case you take my advice and remove the double-documented typing, but if not I'll merge it anyway; it's not hurting anything.

File: .github/actions/pr-to-update-go/pr_to_update_go/go_pr_maker.py (hunk 4 above)
Comment (zrhoffman) [reply .github/actions/pr-to-update-go/pr_to_update_go/go_pr_maker.py]: Removed typing docstrings in 4eccc0bd80.

File: .github/actions/build-rpms/build-rpms.sh (hunk 3 above)
Comment (ocket8888) [reply .github/actions/build-rpms/build-rpms.sh:38]: no, that's fine

**Review by ocket8888 — APPROVED** [review]

**Pull request merged**